src/inc/til/u8u16convert.h

```diff
@@ -383,17 +305,83 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
     // - state - reference to a til::u16state class holding the status of the current partials handling
     // Return Value:
     // - S_OK          - the conversion succeeded without any change of the represented code points
+    // - S_FALSE       - the resulting string is converted from the previously cached partial only
     // - E_OUTOFMEMORY - the function failed to allocate memory for the resulting string
     // - E_ABORT       - the resulting string length would exceed the upper boundary of an int and thus, the conversion was aborted before the conversion has been completed
     // - E_UNEXPECTED  - an unexpected error occurred
-    template<class inT, class outT>
-    [[nodiscard]] typename std::enable_if<std::is_same<typename inT::value_type, wchar_t>::value && std::is_same<typename outT::value_type, char>::value, HRESULT>::type
-    u16u8(const inT in, outT& out, u16state& state) noexcept
+    template<class outT>
+    [[nodiscard]] HRESULT u16u8(const std::wstring_view& in, outT& out, u16state& state) noexcept
     {
-        std::wstring_view sv{};
-        RETURN_IF_FAILED(state(std::wstring_view{ in }, sv));
-        return u16u8(sv, out);
+        try
+        {
+            int len16{};
+            int capa8{};
+            // The worst ratio of UTF-16 code units to UTF-8 code units is 1 to 3.
+            RETURN_HR_IF(E_ABORT, !base::MakeCheckedNum(in.length()).AssignIfValid(&len16) || !base::CheckAdd(len16, state.partials[1]).AssignIfValid(&capa8) || !base::CheckMul(capa8, 3).AssignIfValid(&capa8));
+
+            out.clear();
+            RETURN_HR_IF(S_OK, !capa8); // nothing to convert, however this is not an error
+
+            out.resize(gsl::narrow_cast<size_t>(capa8));
+            int len8{};
+            auto cursor16{ in.data() };
+            if (!!state.partials[1])
+            {
+                if (!len16) // only the partial code point will be converted and published

```

Comment:
This is always false.

Reply:
Ihave difficulties to understand why 🤔
![image](https://user-images.githubusercontent.com/46659645/129803874-6c3b8130-d31f-46f5-90ea-522c52776bff.png)
Dosn't this prove that we still could have partials in the cache even if we got a zero-length string

Reply:
`capa8` is `(len16 + state.partials[1]) * 3` and you return early if `capa8` is 0.
`capa8` cannot be 0 if `len16` is 0. That's why I believe this condition cannot ever be true. In fact I'd say we should always return early if `len16` (or the corresponding `len8` in the other function) is 0, since that implies that the input buffer is empty.

Reply:
I'm of two minds what to do with the bytes we might still have in the cache. In my screenshot above it's still converted (the replacement character at the beginning of the third line). If we decide to return an empty string we have two possibilities
* reset the cache and discard the partials
* leave the cache as is and wait for any subsequent call that gives us the missing bytes to complete the code point.

What do you think is the correct way.

Reply:
@german-one I strongly believe the second case is the correct one. In fact it's how it's been implemented so far, including in your new implementation. And it makes sense for me: If the input data is empty no transformations are needed. Finally the promise of u8u16 with state is that it emits only full/paired code points. If we discarded partials, because u8u16 was called with an empty string, then we'd kinda violate that promise.

Reply:
That's absolutely logical.